src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -1203,8 +1205,73 @@ public bool TryFormat(Span<char> destination, out int charsWritten, [StringSynta
                         p += HexsToCharsHexOutput(p, _j, _k);
                         *p++ = '}';
                     }
+                    else if ((Ssse3.IsSupported || AdvSimd.Arm64.IsSupported) && BitConverter.IsLittleEndian)
```

Comment:
Is having all three decently large implementations (hex, simd, fallback) going to be problematic for the JIT?

We've placed them in helpers previously to ensure that the main dispatch check is inlined and the core logic gets the call instead.

Reply:
I've checked - it's inline without issues. And the `TryFormat` method itself won't be inlined anywhere (as expected).

Reply:
This is already a large method. We wouldn't want or expect it to be inlined with or without the vectorization, would we?

Reply:
Right, it should not even pass the quick IL size check (128 bytes of IL without PGO)

Reply:
> This is already a large method. We wouldn't want or expect it to be inlined with or without the vectorization, would we?

This depends and is what I was asking around. 

`TryFormat` is large, but it is effectively 5 methods rolled into 1. Given that which of the 5 methods is used is dependent on `format` which is often a "constant", then it might be desirable to refactor this code a bit...

Specifically, right now we have:
```
// arg checks

// per format control flow, which is constant per format

if (hex)
{
    // hex code
}
else
{
    // fallback code
}
```

This results in a very large method that is `1330` bytes long and that the JIT simply refuses to inline.

Refactoring the code a bit can result in a smaller and inlinable (when constant format is passed in) `TryFormat` which does the initial dispatch logic and then defers to the actual "core logic" (which remains a call).

----

For example, today this code:
```csharp
public static void Test(MyGuid guid, Span<char> destination)
{
    guid.TryFormat(destination, out _, default);
}
```

Compiles down to roughly:
```asm
; Method C:Test(MyGuid,System.Span`1[ushort])
G_M65252_IG01:              ;; offset=0000H
       4883EC38             sub      rsp, 56
       C5D857E4             vxorps   xmm4, xmm4
       C5F97F642420         vmovdqa  xmmword ptr [rsp+20H], xmm4
       33C0                 xor      eax, eax
       4889442430           mov      qword ptr [rsp+30H], rax
						;; size=21 bbWeight=1 PerfScore 3.83

G_M65252_IG02:              ;; offset=0015H
       4533C9               xor      r9, r9
       4C894C2420           mov      bword ptr [rsp+20H], r9
       4533C9               xor      r9d, r9d
       44894C2428           mov      dword ptr [rsp+28H], r9d
       4C8D4C2420           lea      r9, [rsp+20H]
       4C8D442430           lea      r8, [rsp+30H]
       FF15FBCF8200         call     [MyGuid:TryFormat(System.Span`1[ushort],byref,System.ReadOnlySpan`1[ushort]):bool:this]
       90                   nop      
						;; size=33 bbWeight=1 PerfScore 6.75

G_M65252_IG03:              ;; offset=0036H
       4883C438             add      rsp, 56
       C3                   ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 59
```

Tweaking things a bit allows this instead:
```asm
; Method C:Test(MyGuid,System.Span`1[ushort])
G_M65252_IG01:              ;; offset=0000H
       4883EC48             sub      rsp, 72
       C5D857E4             vxorps   xmm4, xmm4
       C5F97F642430         vmovdqa  xmmword ptr [rsp+30H], xmm4
       33C0                 xor      eax, eax
       4889442440           mov      qword ptr [rsp+40H], rax
						;; size=21 bbWeight=1 PerfScore 3.83

G_M65252_IG02:              ;; offset=0015H
       4C8B02               mov      r8, bword ptr [rdx]
       8B5208               mov      edx, dword ptr [rdx+08H]
       4C89442430           mov      bword ptr [rsp+30H], r8
       89542438             mov      dword ptr [rsp+38H], edx
       33D2                 xor      edx, edx
       89542420             mov      dword ptr [rsp+20H], edx
       C744242801000000     mov      dword ptr [rsp+28H], 1
       488D542430           lea      rdx, [rsp+30H]
       4C8D442440           lea      r8, [rsp+40H]
       41B924000000         mov      r9d, 36
       FF1560D08200         call     [MyGuid:<TryFormat>g__Fallback|12_1(byref,System.Span`1[ushort],byref,int,int,bool):bool]
       90                   nop      
						;; size=52 bbWeight=1 PerfScore 12.75

G_M65252_IG03:              ;; offset=0049H
       4883C448             add      rsp, 72
       C3                   ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 78
```

This is only slightly larger (and could actually be improved a bit on the JIT side of things) but skips all the unnecessary "cruft" that otherwise exists and is unnecessary for the common case.

Reply:
This is, for example, exactly why many of our vectorized paths do:
```
if (Vector128.IsHardwareAccelerated && (array.Length >= Vector128<T>.Count))
{
    VectorImpl(....);
}
else
{
    FallbackImpl(...);
}
```

Because this allows the non-expensive "dispatch logic" (the size check) to be inlined into the caller while keeping the more expensive core logic in a call.

It notably also means the JIT sees smaller IL bodies which can allow it to more easily optimize the code and do downstream inlining for the two Impl paths.

Reply:
I see what you mean, that makes sense, although, it might hit the same inlining limitation we discussed the other day:
if you make TryFormat inlineable - it won't longer be the root of call graph so some important nested parts of TryFormat might not have enough budget to inline (e.g. `HexsToChars`)

I'll experiment with this separately if you don't mind. The idea of making TryFormat potentially inlineable definitely makes sense, considering that TryFormat is often used with a constant format letter

Reply:
> This is, for example, exactly why many of our vectorized paths do:

Here it's a bit different pattern:

<preparations>
if (format == hex)
{
    // a lot of stuff
}
else if (SIMD)
{

}
else
{
    // fallback
}

Reply:
it's easy to do if we can pessimize the hex format, but it needs to be confirmed that the format is rarely-used (and not on hot paths).

Reply:
> Tweaking things

This was my main point. We could restructure things, but _just_ putting the vectorized dbnp and fallback dbnp paths into separate methods is insufficient to make TryFormat something we'd want to always inline. If we want to restructure this in a follow-up, that's fine.